packages/core/src/RenderPipeline/BasicRenderPipeline.ts

```diff
@@ -77,19 +72,21 @@ export class BasicRenderPipeline {
     const sunlight = scene._lightManager._sunlight;
     const depthOnlyPass = this._depthOnlyPass;
     const depthPassEnabled = camera.depthTextureMode === DepthTextureMode.PrePass && depthOnlyPass._supportDepthTexture;
-    camera.engine._spriteMaskManager.clear();
 
     if (scene.castShadows && sunlight && sunlight.shadowType !== ShadowType.None) {
       this._cascadedShadowCasterPass.onRender(context);
     }
 
+    const batcherManager = engine._batcherManager;
+    const maskManager = engine._maskManager;
     cullingResults.reset();
-    this._allSpriteMasks.length = 0;
+    maskManager.clear();
```

Comment:
There should be two steps here：
- collect mask: should before `MaskRenderer` commit render data.
- reset preMaskLayer: better to put it in `_drawRenderPass`, because it is closely related to the screen's stencil buffer.

```
    if (clearFlags !== CameraClearFlags.Stencil) {
      maskManager._preMaskLayer = 0;
    }
```

Reply:
done